crmsh/corosync.py

```diff
@@ -669,215 +649,370 @@ def remove_qdevice_db(self):
+        return re.search("{}__[0-9]+$".format(path), inner_key)
 
-def make_value(path, value):
-    "Create a token sequence representing a value"
-    sp = path.split('.')
-    name = sp[-1]
-    return [Token(_tVALUE, sp, name, value)]
+    @staticmethod
+    def _matched_inner_section_end(path, inner_sec_end):
+        """
+        Check if given path matched an inner end section
+        """
+        # example of inner section end:
+        #   totem.interface__0__END
+        #   nodelist.node__1__END
+        #   quorum__0__END
+        return re.search("{}__[0-9]+__END$".format(path), inner_sec_end)
 
+    @staticmethod
+    def _inner_section_name(inner_key):
+        """
+        Extract inner section name from an inner key
+        """
+        # nodelist.node.nodeid__1 -> nodelist.node__1
+        return re.sub("(.*)\.\w+(__[0-9])", "\\1\\2", inner_key)
 
-class Parser(object):
-    def __init__(self, data):
-        self._tokens = list(corosync_tokenizer(data))
-
-    def find(self, name, start=0):
-        """Gets the index of the element with the given path"""
-        for i, t in enumerate(self._tokens[start:]):
-            if t.path == name:
-                return i + start
-        return -1
-
-    def find_bounds(self, name, start=0):
-        """find the (start, end) of the next instance of name found at start"""
-        i = self.find(name, start)
-        if i < 0:
-            return -1, -1
-        if self._tokens[i].token != _tBEGIN:
-            return i, i
-        e = i + 1
-        depth = 0
-        while e < len(self._tokens):
-            t = self._tokens[e]
-            if t.token == _tBEGIN:
-                depth += 1
-            if t.token == _tEND:
-                depth -= 1
-            if depth < 0:
-                break
-            e += 1
-        if e == len(self._tokens):
-            raise ValueError("Unclosed section")
-        return i, e
-
-    def get(self, path):
-        """Gets the value for the key (if any)"""
-        for t in self._tokens:
-            if t.token == _tVALUE and t.path == path:
-                return t.value
-        return None
+    @staticmethod
+    def _section_name(inner_key):
+        """
+        Extract section name from an inner key
+        """
+        # nodelist.node.nodeid__1 -> nodelist.node
+        return re.sub("(.*)\.\w+(__[0-9])", "\\1", inner_key)
 
-    def get_all(self, path):
-        """Returns all values matching path"""
-        ret = []
-        for t in self._tokens:
-            if t.token == _tVALUE and t.path == path:
-                ret.append(t.value)
-        return ret
+    @staticmethod
+    def _key_name(inner_key):
+        """
+        Extract key name from an inner key
+        """
+        # nodelist.node.nodeid__1 -> nodelist.node.nodeid
+        return re.sub("__[0-9]+", "", inner_key)
 
-    def all_paths(self):
-        """Returns all value paths"""
+    @property
+    def _inner_section_name_list(self):
```

Comment:
I'm not sure if it make sense to define a property which is basically "pseudo-private". Usually, in most cases, properties should be public.
If you want to make this property public, remove the underscore in front of the function name.

Reply:
No, I don't want it to be public, like its name, it was used internally